Spherical inclusions

arjunramaswami opened the pull request:
Implemented an example to simulate the response of a matrix material with spherical inclusions in which the FFT operations can also be offloaded to the GPU with the help of CuPy

arjunramaswami committed:
Added small strain spherical example with GPU functionality

arjunramaswami committed:
Modified Readme with application description

arjunramaswami committed:
Minor correction in README link

ahenkes1 committed:
Update linear_elasticity_even.py

Small fix in a comment

ahenkes1 committed:
Update README.md

arjunramaswami committed:
Updated Readme with publication link

tdegeus commented:
Amazing @arjunramaswami and @ahenkes1 thank a lot and very interesting. I'll review asap!

tdegeus commented on small-strain/spherical-inclusions/linear_elasticity_even.py:
```diff
@@ -0,0 +1,156 @@
+import scipy.sparse.linalg as sp
+import itertools
+
+# Activate (1) GPU computation or not (0)
+GPU = 1
+if GPU == 1:
+    import cupy as cp
```
```suggestion
# True: use GPU computations. False: regular CPU computations
GPU = True
if GPU == False:
    import cupy as cp
```

tdegeus commented on small-strain/spherical-inclusions/linear_elasticity_even.py:
```diff
@@ -0,0 +1,156 @@
+
+ndim   = 3            # number of dimensions
+N      = 32           # number of voxels (assumed equal for all directions)
+ndof   = ndim**2*N**3 # number of degrees-of-freedom
+shape  = [N,N,N]  # number of voxels as list: [Nx,Ny,Nz]
```
```suggestion
ndof   = ndim**2*N**3 # number of degrees-of-freedom
shape  = [N,N,N]      # number of voxels as list: [Nx,Ny,Nz]
```

tdegeus commented on small-strain/spherical-inclusions/linear_elasticity_even.py:
```diff
@@ -0,0 +1,156 @@
+for i, j, l, m in itertools.product(range(3), repeat=4):
+    Ghat4[i,j,l,m] = norm*delta(i,m)*q[j]*q[l]
+
+# Activate (1) GPU computation or not (0)
+GPU = 1
+if GPU == 1:
```
```suggestion
if GPU:
```

tdegeus commented on small-strain/spherical-inclusions/linear_elasticity_even.py:
```diff
@@ -0,0 +1,156 @@
+
+# ------------------- PROBLEM DEFINITION / CONSTITIVE MODEL -------------------
+
+# phase indicator: spherical inclusions of volume fraction cf
```
```suggestion
# phase indicator: "nspheres" random spherical inclusions, with total volume fraction "cf"
```

tdegeus commented on small-strain/spherical-inclusions/linear_elasticity_even.py:
```diff
@@ -0,0 +1,156 @@
+
+# phase indicator: spherical inclusions of volume fraction cf
+phase = np.zeros((N, N, N), dtype=np.int)
+real_cf = np.array([phase[phase == 1].size / phase.size])
```
That makes no sense here, `phase` just contains zeros

tdegeus commented on small-strain/spherical-inclusions/linear_elasticity_even.py:
```diff
@@ -0,0 +1,156 @@
+phase = np.zeros((N, N, N), dtype=np.int)
+real_cf = np.array([phase[phase == 1].size / phase.size])
+cf = 0.25
+no_of_spheres = 0
```
```suggestion
nspheres = 0
```

tdegeus commented on small-strain/spherical-inclusions/linear_elasticity_even.py:
```diff
@@ -0,0 +1,156 @@
+if GPU == 1:
+    import cupy as cp
+
+def create_bin_sphere(matrix_size, center, radius):
```
```suggestion
# function to create a sphere (defined by "center" and "radius") 
# as ones in a grid (defined by "shape") of zeros
def create_bin_sphere(shape, center, radius):
```

tdegeus commented on small-strain/spherical-inclusions/linear_elasticity_even.py:
```diff
@@ -0,0 +1,156 @@
+no_of_spheres = 0
+max_r = 0
+min_r = N
+print("\n\n+--------------------------------------------+")
+print("| %-26s %15s |" % (
+    'Number of inclusions', 'Volume fraction'))
+print("+--------------------------------------------+")
```
```suggestion
```
Can be left up to the user, the important thing is to understand the code

tdegeus commented on small-strain/spherical-inclusions/linear_elasticity_even.py:
```diff
@@ -0,0 +1,156 @@
+    'Number of inclusions', 'Volume fraction'))
+print("+--------------------------------------------+")
+while (phase[phase == 1].size / phase.size) <= cf:
+    r = np.random.uniform(low=1, high=10, size=None)
+    if r > max_r:
+        max_r = r
+    if r < min_r:
+        min_r = r
```
To be honest I'm not entirely sure what is going on here

tdegeus commented on small-strain/spherical-inclusions/linear_elasticity_even.py:
```diff
@@ -0,0 +1,156 @@
+        phase[sphere == 1] = 1
+        no_of_spheres += 1
+        real_cf = np.array([phase[phase == 1].size / phase.size])
+        print("| %-20d  %20.3f |" % (no_of_spheres, float(real_cf)))
+print("+--------------------------------------------+\n")
```
```suggestion
```
(same as above)

tdegeus reviewed (changes requested):
Thanks a lot for this example! It is really nice to see how easy it is to use GPUs, and I'm sure it will be appreciated by the community. Furthermore, it is very nice that you took the trouble to vectorize the definition of the projection operator (and to comment it well!).

I made some comments. Beyond them my only concern is the definition of the grid of inclusions. First, I did not understand every detail (see comment), but second, and most important, I think that it should be entirely moved to a function (on top like you did for `create_bin_sphere`). The reason say this is that it is then even easier to read the red-thread, and concern oneself with the problem specific details later. Furthermore, it gives a nice way to split the numerical value of the parameters from the implementation. 

Let me know!!

cc @vondrejc

ahenkes1 committed:
Update small-strain/spherical-inclusions/linear_elasticity_even.py

Co-authored-by: Tom de Geus <tdegeus@users.noreply.github.com>

ahenkes1 committed:
Update small-strain/spherical-inclusions/linear_elasticity_even.py

Co-authored-by: Tom de Geus <tdegeus@users.noreply.github.com>

ahenkes1 committed:
Update small-strain/spherical-inclusions/linear_elasticity_even.py

Co-authored-by: Tom de Geus <tdegeus@users.noreply.github.com>

ahenkes1 committed:
Update small-strain/spherical-inclusions/linear_elasticity_even.py

Co-authored-by: Tom de Geus <tdegeus@users.noreply.github.com>

ahenkes1 committed:
Update small-strain/spherical-inclusions/linear_elasticity_even.py

Co-authored-by: Tom de Geus <tdegeus@users.noreply.github.com>

ahenkes1 committed:
Update small-strain/spherical-inclusions/linear_elasticity_even.py

Co-authored-by: Tom de Geus <tdegeus@users.noreply.github.com>

ahenkes1 committed:
Update small-strain/spherical-inclusions/linear_elasticity_even.py

Co-authored-by: Tom de Geus <tdegeus@users.noreply.github.com>

ahenkes1 commented:
@tdegeus by the way, the vectorization of the projection operator was not done by us, I think credits go to @vondrejc  ;)

ahenkes1 committed:
Update small-strain/spherical-inclusions/linear_elasticity_even.py

Co-authored-by: Tom de Geus <tdegeus@users.noreply.github.com>

ahenkes1 committed:
Implemented function for microstructure creation.

- The creation of the microstructure (spherical inclusions) was debugged
  and moved to a function.

- Minor bug fixes.

arjunramaswami commented:
HI @tdegeus, are there some other changes required from us for the pull request?

tdegeus commented:
Sorry @arjunramaswami . Thanks a lot for your contribution it is highly appreciated! Good luck with the research you do with the code!
